### PR TITLE
PostgreSQL: Only query the sequences and tables which were granted

### PR DESCRIPTION
**Pull request by ChloePlanet**
current database user can only query the sequences and tables

### CODE REVIEW

File: postgresql.go [hunk 1]
```diff
@@ -62,13 +62,11 @@ func (h *PostgreSQL) tableNames(q queryable) ([]string, error) {
 	var tables []string
 
 	sql := `
-	         SELECT pg_namespace.nspname || '.' || pg_class.relname
-		 FROM pg_class
-		 INNER JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
+	   SELECT DISTINCT g.table_schema || '.' || pg_class.relname
```
Comment (andreynering) [inline postgresql.go:65]: Changed indentation. Also, is `DISTINCT` necessary here?

File: postgresql.go [hunk 2]
```diff
@@ -62,13 +62,11 @@ func (h *PostgreSQL) tableNames(q queryable) ([]string, error) {
...
+		 FROM pg_class, information_schema.role_table_grants g
 		 WHERE pg_class.relkind = 'r'
-		 AND
-		 pg_namespace.nspname NOT IN ('pg_catalog', 'information_schema')
-		 AND pg_namespace.nspname NOT LIKE 'pg_toast%';
+		 AND pg_class.relname = g.table_name
+     AND g.table_schema NOT IN ('pg_catalog', 'information_schema');
```
Comment (andreynering) [inline postgresql.go:69]: Changed indentation

File: postgresql.go [hunk 3]
```diff
@@ -62,13 +62,11 @@ func (h *PostgreSQL) tableNames(q queryable) ([]string, error) {
 	var tables []string
 
 	sql := `
-	         SELECT pg_namespace.nspname || '.' || pg_class.relname
-		 FROM pg_class
-		 INNER JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
+	   SELECT DISTINCT g.table_schema || '.' || pg_class.relname
+		 FROM pg_class, information_schema.role_table_grants g
```
Comment (andreynering) [inline postgresql.go:66]: I'd prefer to have a `JOIN` if possible, instead of multiple tables on `WHERE`.

File: postgresql.go [hunk 4]
```diff
@@ -91,10 +89,10 @@ func (h *PostgreSQL) tableNames(q queryable) ([]string, error) {
 
 func (h *PostgreSQL) getSequences(q queryable) ([]string, error) {
 	const sql = `
-		SELECT pg_namespace.nspname || '.' || pg_class.relname AS sequence_name
-		FROM pg_class
-		INNER JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
+		SELECT g.sequence_schema || '.' || pg_class.relname AS sequence_name
+		FROM pg_class, information_schema.sequences g
```
Comment (andreynering) [inline postgresql.go:93]: Same here, can we use a `JOIN` instead of multiple tables on `WHERE`?

**Review by andreynering — COMMENTED** [review]
@CikeyPlanet Thanks for your patch! 🙂 

There are minor formatting issues with the PR.

Also, would be helpful to have more context on the problem.

File: postgresql.go (hunk 1 above)
Comment (ChloePlanet) [reply postgresql.go:65]: Sorry for the wrong indentation, there's something wrong with my new code editor...

About the `DISTINCT`, the sql query statement can be replaced as 
```sql
SELECT g.table_schema || '.' || pg_class.relname FROM pg_class JOIN information_schema.role_table_grants g ON pg_class.relname = g.table_name WHERE pg_class.relkind = 'r' AND g.table_schema NOT IN ('pg_catalog', 'information_schema') AND g.privilege_type = 'SELECT';
```

But my concern is that postgresql's fine-grained grants(SELECT/INSERT/UPDATE/DELETE/TRUNCATE etc.). If we just check the `SELECT` permission when `Load()`, things may go wrong later. Do you have some ideas how to trade off this thing?

**Comment by andreynering** [comment]
@CikeyPlanet Sorry for the long time to respond.

What do you think about?

```sql
SELECT pg_namespace.nspname || '.' || pg_class.relname
FROM pg_class
INNER JOIN pg_namespace ON pg_namespace.oid = pg_class.relnamespace
WHERE pg_class.relkind = 'r'
  AND pg_namespace.nspname NOT IN ('pg_catalog', 'information_schema')
  AND pg_namespace.nspname NOT LIKE 'pg_toast%'
  AND EXISTS (
    SELECT 1
    FROM information_schema.role_table_grants
    WHERE role_table_grants.grantee = current_user
      AND role_table_grants.table_schema = pg_namespace.nspname
      AND role_table_grants.table_name = pg_class.relname
  )
ORDER BY 1
```

(I haven't tried adapting the sequences one).

As a workaround while this is not merged, you can use `NewFiles` instead of `NewFolder`.

**Comment by ChloePlanet** [comment]
@andreynering This one worked for me:)

In addition, `NewFiles` isn't okay cause the code we changed which in function `tableNames` will be called in `init()`.

**Comment by andreynering** [comment]
@CikeyPlanet Any updates on this patch?

**Comment by andreynering** [comment]
Closing for lack of feedback

**Pull request closed without merging**